design-docs/007-basic-expiration.md

```diff
@@ -0,0 +1,561 @@
+is where the storage savings actually happen, by deleting chunks that were
+overwritten (or flagged as deleted) in later versions of the repository.
+
+Keeping history is still very useful: it provides the infrastructure for
+transactions, it allows users to undo mistakes, improves auditability, and,
+in certain cases, it's a core feature users need (example: insurance industry).
+
+
+## The simpler problem of snapshot removal
+
+To understand how to expire old snapshots, we first need to understand
+how we can abandon a series of arbitrary snapshots.
+
+We present a diagram of a repository snapshots. Arrows follow the `parent`
+relationship. Above in the diagram is the original repo, and below, we show
+two possible ways to abandon snapshots `X1, X2, ..., Xn`
+
+```mermaid
+stateDiagram-v2
+  direction TB
+  state Original {
+
+    state "X1, X2, ..., Xn" as X
+    state "..." as REST
+
+    [*] --> N1
+    N1 --> N2
+    N2 --> A
+    A --> X
+    X --> B
+    B --> REST
+    
+  }
+
+  Original --> Skip
+  Original --> EditParents
+
+  state "Update A in place" as Skip {
+
+    state "N1" as N1E
+    state "N2" as N2E
+    state "A (updated)" as AE
+    state "B" as BE
+    state "X1, X2, ..., Xn" as XE
+    state "..." as RESTE   %% codespell:ignore
+
+    [*] --> N1E
+    N1E --> N2E
+    N2E --> AE
+    AE --> BE
+    XE --> BE
+    BE --> RESTE    %% codespell:ignore
+
+  }
+
+  state "New A" as EditParents {
+    state "N1'" as N1P
+    state "N2'" as N2P
+    state "A'" as AP
+    state "N1" as N1POLD
+    state "N2" as N2POLD
+    state "A" as APOLD
+    state "B"  as BP
+    state "X1, X2, ..., Xn" as XP
+    state "..." as RESTP     %% codespell:ignore
+
+    [*] --> N1P
+    N1POLD --> N2POLD
+    N2POLD --> APOLD
+    APOLD --> XP
+    XP --> BP
+    N1P --> N2P
+    N2P --> AP
+    AP --> BP
+    BP --> RESTP   %% codespell:ignore
+  }
+```
+In this diagram we want
+to abandon a stream of contiguous snapshots `X1, X2,..., Xn` that seats
+in the middle of some branch's history.
+
+There are many ways in which this could be achieved. We display in the
+diagram and explain below, the two we think deserve more analysis.
+
+### Left algorithm: Update `A` in place
+
+On the alternative displayed on the left of the diagram, we first find
+snapshot `A`. It is found by traversing the commits from the
+branch/tag pointer, using the `parent` relationship, and stopping
+right before the first snapshot we want to abandon.
+
+Once we have found `A` we can update its contents in place, maintaining its `SnapshotId`,
+so, maintaining its path in object store. The update we do is
+changing its ancestry, keeping all ancestors except the set
+`{X1, X2, ..., Xn}`. In this form, `A`'s direct parent becomes `B`.
+
+No other changes are needed. Snapshots `X1, X2,..., Xn` remain on
+disk, they are just no longer accessible from the starting branch/tag.
+
+### Right algorithm: New `A`
```

Comment:
why not combine both? create `N1 -> N2 -> A' -> B'` so a new A' is created but parents are unaffected, only descendants are rewritten -- this is squash-rebase in git. that way any references to `N1` are preserves (e.g. a deployed app).

Reply:
If I modify `A` to an `A'` I need to modify `N2` because it has to point to its parent `A'`, then `N1`, etc